### ISSUE TITLE
Request - Don't show next line of lyrics so soon after line breaks

### ISSUE THREAD

**Issue by AlecScudder** [Vocaluxe/Vocaluxe#255]
I am modifying TXT files to emulate my favorite mode from SingStar - Co Op.  This mode lets two users trade off singing the same vocal track (P1 sings a couple lines, then P2 sings a couple lines, then they both sing together, etc.). Everyone loves this mode at parties because they get lots of "solos" and get to find their pitch with the main song only (instead of off key buddies always singing). But it also pits both players head to head at times, too. Best of both worlds!

So I modify the TXT files that are out there into P1 and P2 sections accordingly (repeating the lines they both sing in both sections of the file, and keeping their solo parts in their respective sections). 

However, when only 1 player is singing, the other player sees their next line's song lyric come up WAY before they are going to sing it.  This can be very confusing when trying to play this DUET / CO-OP mode. Lyrics often repeat and players are getting confused on when it is really their turn to sing because they are seeing their players lyrics come up while they still have to wait for the other player to sing 2 or 3 solo lines.

I've tried adjusting the line line break numbers to accommodate this (e.g. to try to delay showing the player's next line until a couple seconds before it is supposed to be sung) but the line break numbers seems to have no effect on when the lyrics show.

**Comment by AlecScudder** [comment]
By the way, meant to say how much I LOVE this project!!! I just discovered it and I cannot tell you how much joy this is going to bring to me and my friends - particularly if we can find a way to fix this issue. Thanks!!!

**Comment by flokuep** [comment]
Thanks :)

I guess it should work as you expect it. Lyrics and notes fade in and out when there are bigger breaks. Which version of vocaluxe do you use? If it's 0.2.1, please test our latest nightly or beta version. I've currently tested it by myself and don't find any problem.

Side note: Maybe it would be a nice feature to do some automatic co op songs? 🤔

**Comment by AlecScudder** [comment]
Yes, using 0.2.1.  I will try nightly build.  However, can you tell how long a "bigger" break is? Is it driven by seconds? Number of lines int he text file? Something with the numbers or line breaks? If I knew how long the break needed to be I could adjust the TXT file accordingly. But right now I'm guessing at how to break P1 and P2 to allow that fade out of lyrics and notes which helps this co-op mode tremendously! 

If it is unreasonable to reduce the threshold of what defines a "bigger" break, at least if I knew how to calculate the minimum break needed I could adjust the files accordingly to most closely emulate the co op experience. 

And yes and automatic co op might be nice, however, it would seem hard to do because you might end up giving all the "good" parts to one player or the other. But I'd love to see it and try it out.

Thanks!

**Comment by flokuep** [comment]
We have a setting in our config.xml for this :) (I guess this was added sometimes after 0.2.1)
`
  <MinLineBreakTime>0.1</MinLineBreakTime>`

We're currently trying to find refrains of songs to get some nice medley parts... Could reuse this maybe...

**Comment by AlecScudder** [comment]
Thanks for the response. I tried the nightly build but have the same issue. The minimum line break time is actually the _opposite_ of what would be needed. If there was a MAXIMUM line break time (maximum seconds the text/notes are shown before they are to be sung) that would be exactly what I needed. I could set that to 2 or 3 seconds and it would be perfect!

**Comment by AlecScudder** [comment]
I realize with more testing also why it is so distracting in this mode. This is what often happens...

P1 is singing, P2 is inactive for a while so their lyrics and notes are gone. P1 still has 3 or 4 more lines to sing before it switches to the lines that P2 will sing, BUT P2's lyrics/notes show up so early that both players tend to get distracted and are not sure which line to sing.

Any method you come up with to help solve would be much appreciated.

**Comment by AlecScudder** [comment]
I was able to modify the following line of code from CScreenSing to get it close to what I need. This makes the lyrics disappear faster and more frequently so it is more apparent when the other singer has solo lines. This is from CSCreenSing:

 private float[] _CalcFadingAlpha()
        {
            const float dt = 1.5f;

**Comment by flokuep** [comment]
Could you send me an example text file so I could check it by msyself (maybe with pastebin or something else...)?

**Comment by AlecScudder** [comment]
Here is the pastebin of Code Monkey as CO-OP.   http://pastebin.com/uaX9cBGG

In this example, P1 sings a bit, then P2, then both.  There is a 13 second gap between when the first P1 vocal ends and the next time P1 sings (at same time as P1). BUT... P1's lyrics / notes never really fade out and their next lyrics / note is displayed a full 10+ seconds before they are actually supposed to sing it - even though P2 is singing completely different lyrics in the song during that 10-13 second gap.  

Many songs have repetitive / similar lyrics throughout so it becomes confusing for both players unless the person NOT singing has their lyrics fade out more quickly (and / or fade in later).

If you run it hopefully you can see how this is confusing. Making the code change above I mentioned shortens the time to fade out and makes it a bit less confusing because the lyrics fade quicker. However, ideally the new lyrics should not appear as quickly, either... maybe just a few seconds before they need to be sung.

**Comment by lukeIam** [comment]
> const float dt = 1.5f;

I had a look on the code and I agree that's the right variable to adjust the point where the line is shown.  
But keep in mind that it is also influencing the fade out point.  
Would it be an option to make the `dt` option configurable in the config file?

The code was not written by me but here is a quick analysis of the fade in logic (note: alpha = dt && diff=next):
 
![image](https://cloud.githubusercontent.com/assets/5115160/21775666/4c5edc36-d697-11e6-8e85-8e03ddf89cae.png)

![image](https://cloud.githubusercontent.com/assets/5115160/21775404/372e6634-d696-11e6-9206-99fbbfa3ace6.png)

**Comment by AlecScudder** [comment]
Yes, making the dt option configurable in the config file would be ideal!
